SQL: Add xDBC and CLI support. QA CSV specs

bpintea opened the pull request:
This adds support for the xDBC and CLI clients.
Based on these, the CsvJdbc-based QA tests are also added.

bpintea committed:
Add xDBC and CLI support. QA CSV specs

This adds support for the xDBC and CLI clients.
Based on those, the CsvJdbc-based QA tests are also integrated.

bpintea committed:
Improve array testing

Adds a REST test that generates all possible path combination of a JSON
object, all generating values to be retrieved with ARRAY().

bpintea committed:
Merge remote-tracking branch 'upstream/feat/sql-multivalue' into feat/multivalue-clients

bpintea commented:
@elasticmachine run elasticsearch-ci/2

bpintea commented:
@elasticmachine run elasticsearch-ci/2

bpintea committed:
Merge remote-tracking branch 'upstream/feat/sql-multivalue' into feat/multivalue-clients

bpintea committed:
Test fix

Correct timezone formatting pattern.

elasticmachine commented:
Pinging @elastic/es-ql (Team:QL)

matriv commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/stats/VerifierMetricsTests.java:
```diff
@@ -161,7 +161,7 @@ public void testWhereLimitGroupByHavingOrderByQuery() {
     public void testTwoQueriesExecuted() {
         Metrics metrics = new Metrics();
-        Verifier verifier = new Verifier(metrics);
+        Verifier verifier = new Verifier(metrics, SqlTestUtils.TEST_CFG.version());
```
static import?

matriv commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryFolderTests.java:
```diff
@@ -53,7 +53,8 @@ public static void init() {
         EsIndex test = new EsIndex("test", mapping);
         IndexResolution getIndexResult = IndexResolution.valid(test);
-        analyzer = new Analyzer(SqlTestUtils.TEST_CFG, new SqlFunctionRegistry(), getIndexResult, new Verifier(new Metrics()));
+        analyzer = new Analyzer(SqlTestUtils.TEST_CFG, new SqlFunctionRegistry(), getIndexResult,
```
static import of TEST_CFG?

matriv commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/optimizer/OptimizerRunTests.java:
```diff
@@ -77,7 +77,8 @@ public OptimizerRunTests() {
         EsIndex test = new EsIndex("test", mapping);
         getIndexResult = IndexResolution.valid(test);
-        analyzer = new Analyzer(SqlTestUtils.TEST_CFG, functionRegistry, getIndexResult, new Verifier(new Metrics()));
+        analyzer = new Analyzer(SqlTestUtils.TEST_CFG, functionRegistry, getIndexResult,
```
same here

matriv commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/plan/logical/command/sys/SysTablesTests.java:
```diff
@@ -335,7 +335,7 @@ private SqlTypedParamValue param(Object value) {
         EsIndex test = new EsIndex("test", mapping);
         Analyzer analyzer = new Analyzer(SqlTestUtils.TEST_CFG, new FunctionRegistry(), IndexResolution.valid(test),
-                                         new Verifier(new Metrics()));
+                                         new Verifier(new Metrics(), SqlTestUtils.TEST_CFG.version()));
```
same here

matriv commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java:
```diff
@@ -136,7 +136,7 @@
             EsIndex test = new EsIndex("test", mapping);
             IndexResolution getIndexResult = IndexResolution.valid(test);
-            analyzer = new Analyzer(TEST_CFG, sqlFunctionRegistry, getIndexResult, new Verifier(new Metrics()));
+            analyzer = new Analyzer(TEST_CFG, sqlFunctionRegistry, getIndexResult, new Verifier(new Metrics(), TEST_CFG.version()));
```
same here

matriv commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/plan/logical/command/sys/SysTypesTests.java:
```diff
@@ -142,4 +147,35 @@ public void testSysTypesMultipleMatches() {
+            }
+        }
+    }
+
```
minor: new lines

matriv commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/CurrentTimeTests.java:
```diff
@@ -93,7 +93,8 @@ public void testInvalidPrecision() {
                 SqlTypesTests.loadMapping("mapping-multi-field-with-nested.json")));
 
-        Analyzer analyzer = new Analyzer(SqlTestUtils.TEST_CFG, new SqlFunctionRegistry(), indexResolution, new Verifier(new Metrics()));
+        Analyzer analyzer = new Analyzer(SqlTestUtils.TEST_CFG, new SqlFunctionRegistry(), indexResolution,
```
same

matriv commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/CurrentDateTimeTests.java:
```diff
@@ -92,7 +92,8 @@ public void testInvalidPrecision() {
                 SqlTypesTests.loadMapping("mapping-multi-field-with-nested.json")));
 
-        Analyzer analyzer = new Analyzer(SqlTestUtils.TEST_CFG, new SqlFunctionRegistry(), indexResolution, new Verifier(new Metrics()));
+        Analyzer analyzer = new Analyzer(SqlTestUtils.TEST_CFG, new SqlFunctionRegistry(), indexResolution,
```
same

matriv commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java:
```diff
@@ -51,7 +52,8 @@ private String error(String sql) {
 
     private String error(IndexResolution getIndexResult, String sql) {
-        Analyzer analyzer = new Analyzer(TEST_CFG, new SqlFunctionRegistry(), getIndexResult, new Verifier(new Metrics()));
+        Analyzer analyzer = new Analyzer(TEST_CFG, new SqlFunctionRegistry(), getIndexResult, new Verifier(new Metrics(),
```
same

matriv commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java:
```diff
@@ -71,7 +73,8 @@ private EsIndex getTestEsIndex() {
 
     private LogicalPlan accept(IndexResolution resolution, String sql) {
-        Analyzer analyzer = new Analyzer(TEST_CFG, new SqlFunctionRegistry(), resolution, new Verifier(new Metrics()));
+        Analyzer analyzer = new Analyzer(TEST_CFG, new SqlFunctionRegistry(), resolution, new Verifier(new Metrics(),
```
same

bpintea replied:
Fixed hopefully everywhere (except here, where it was already statically imported and used)

matriv replied:
Thank you!

matriv commented on x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/DataLoader.java:
```diff
@@ -396,15 +454,58 @@ private static void csvToLines(String name, CheckedBiConsumer<List<String>, List
         }
     }
 
+    private static List<String> splitCsvLine(String line) {
```
Have you checked if `org.relique.jdbc.csv` (which we use already) has this functionality?

bpintea replied:
They do have the functionality, but don't export it.

matriv commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/TypeUtils.java:
```diff
@@ -176,4 +176,22 @@ static EsType of(Class<? extends Object> clazz) throws SQLException {
+    }
+
+    static boolean isArray(EsType type) {
+        return type.getName().endsWith("_ARRAY");
```
Couldn't we have a flag on the EsType for that instead of checking the name?

bpintea replied:
Right, thanks, fixed to check against `Types.ARRAY`.

matriv commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/JdbcResultSet.java:
```diff
@@ -934,7 +935,11 @@ public Clob getClob(int columnIndex) throws SQLException {
-        throw new SQLFeatureNotSupportedException("Array not supported");
+        EsType type = columnType(columnIndex);
+        if (isArray(type) == false) {
+            throw new SQLException("Cannot get an array for the column [" + columnIndex + "] of type [" + type.getName() + "]");
```
Maybe `Cannot get column ... of type ... as array`?

bpintea replied:
Better, fixed, thx.

matriv reviewed (commented):
Looks good!! Had a first pass and left some mostly minor comments.

matriv commented:
Can we have some test with prepared statement where an array is set to a query param? (Of course param currently can only be in the SELECT clause).

If it's not in the scope of this PR to have this, then let's have a simple test that tests the Exception thrown.

bpintea committed:
Address review comments

- static imports
- exception message rephrasing
- better array type checking.

bpintea commented:
> Can we have some test with prepared statement where an array is set to a query param? (Of course param currently can only be in the SELECT clause).

Array parameters are indeed out of scope atp (just like literals or constructors).

> If it's not in the scope of this PR to have this, then let's have a simple test that tests the Exception thrown.

Not sure how to go about this:
* passing a JSON array as parameter will trigger a parsing "regular" exception (in the regular request parser); so this is partially covered.
* passing a scalar with an `_array` type (to smth like `SELECT ?`) will simply be ignored. This is the case btw with intervals too. We can add a filter here, I guess, but wondering why hasn't it already been added, to restrict it to known types.

matriv commented:
> * passing a scalar with an `_array` type (to smth like `SELECT ?`) will have the type be simply ignored. This is the case btw with intervals too. We can add a filter here, I guess, but wondering why hasn't it already been added, to restrict it to allowed types.

I meant to have a test for `stmt.setArray()` that currently should throw an exception (unsupported type).

astefan commented on x-pack/plugin/sql/qa/jdbc/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/JdbcTestUtils.java:
```diff
@@ -104,7 +113,22 @@ static int extractNanosOnly(long nanos) {
 
+    static <T> Set<T> randomSet(int minSetSize, int maxSetSize, Supplier<T> valueConstructor) {
+        final int size = randomIntBetween(minSetSize, maxSetSize);
+        Set<T> set = new HashSet<>();
```
Why not putting a `size` to the HashSet constructor, since it's already known. To help with this a bit, we have in CollectionUtils the method `mapSize`.

bpintea replied:
`CollectionUtils` is in `ql`, which isn't a dependency. Not sure if optimising a random-size-driven test would be worth, but still, I could add the method ("formula") here if you think it'd be better that way.

astefan replied:
Not worth adding ql as a dependency for this.

astefan reviewed (changes requested):
Left a first round of reviews, after going through roughly half of the code changes.

bpintea committed:
Address review comments

- fix index bounds check and add tests for it;
- imporove testing;
- minor refactorings.

bpintea commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/JdbcArray.java:
```diff
@@ -0,0 +1,99 @@
+        if (index < 0 || index > Integer.MAX_VALUE) {
+            throw new SQLException("Index value [" + index + "] out of range [0, " + Integer.MAX_VALUE + "]");
+        }
+        int index0 = (int) index - 1; // 0-based index
```
Moved.

bpintea commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/JdbcArray.java:
```diff
@@ -0,0 +1,99 @@
+        if (count < 0) {
+            throw new SQLException("Illegal negative count [" + count + "]");
+        }
+        int available = index0 < values.size() ? values.size() - index0 : 0;
```
> From what I can see index can be 0.

It can not, it is [1-based index](https://docs.oracle.com/en/java/javase/15/docs/api/java.sql/java/sql/Array.html#getArray(long,int)).
The out of range exception you exemplified [below](https://github.com/elastic/elasticsearch/pull/68966#discussion_r576726124) was triggered due to an incorrect initial check, `index < 0`, which is now corrected to `< 1` (I also assumed it's 0-based... as an array, so wrote the check, then confronted with the docs, but missed to fix the check :-/).

bpintea commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/TypeUtils.java:
```diff
@@ -176,4 +177,22 @@ static EsType of(Class<? extends Object> clazz) throws SQLException {
+            : type;
+    }
+
+    static EsType arrayOf(EsType type) throws SQLException {
```
> It is because ENUM_NAME_TO_TYPE is local to TypeUtils?

Yes.

(`TypeUtils#isString()` is btw also only used once, in `JdbcPreparedStatement`, with no `TypeUtils` dependencies.)

bpintea commented on x-pack/plugin/sql/jdbc/src/test/java/org/elasticsearch/xpack/sql/jdbc/JdbcArrayTests.java:
```diff
@@ -0,0 +1,62 @@
+    static final List<EsType> ARRAY_TYPES = Arrays.stream(EsType.values()).filter(TypeUtils::isArray).collect(Collectors.toList());
+
+    public void testMetaData() throws Exception {
+        EsType arrayType = randomFrom(ARRAY_TYPES);
```
Exhaustive-ised.

bpintea commented on x-pack/plugin/sql/jdbc/src/test/java/org/elasticsearch/xpack/sql/jdbc/JdbcArrayTests.java:
```diff
@@ -0,0 +1,62 @@
+        Object[] contained = (Object[]) array.getArray(4, 3);
+        assertEquals(asList(3, 4, 5), asList(contained));
+
+        Object[] overlapping = (Object[]) array.getArray(9, 3);
```
Nice, thanks, fixed.

bpintea commented:
> I meant to have a test for stmt.setArray() that currently should throw an exception (unsupported type).

I see. Makes sense, I've added it.

bpintea committed:
Merge remote-tracking branch 'upstream/feat/sql-multivalue' into feat/multivalue-clients

bpintea committed:
Adress review comments

- more comments
- refactor testing class
- remove unnecessary brackets.

bpintea committed:
Merge remote-tracking branch 'upstream/feat/sql-multivalue' into feat/multivalue-clients

bpintea commented on x-pack/plugin/sql/qa/jdbc/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/ResultSetTestCase.java:
```diff
@@ -180,6 +186,49 @@ public void testMultiValueFields_InsideObjects_WithMultiValueLeniencyDisabled()
+            List<Integer> expected = asList(ints);
+            expected.sort(Integer::compare);
+            List<?> actual = asList((Object[]) intArray.getArray());
+            actual.sort(Comparator.comparing(x -> ((Integer) x)));
```
Applied, thanks.

bpintea commented on x-pack/plugin/sql/qa/jdbc/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/ResultSetTestCase.java:
```diff
@@ -180,6 +186,49 @@ public void testMultiValueFields_InsideObjects_WithMultiValueLeniencyDisabled()
+
+            strings.sort(String::compareTo);
+            actual = asList((Object[]) strArray.getArray());
+            actual.sort(Comparator.comparing(x -> ((String) x)));
```
Fixed.

bpintea commented on x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/DataLoader.java:
```diff
@@ -372,6 +375,61 @@ protected static void loadLibDatasetIntoEs(RestClient client, String index) thro
+            builder.append("}").append("\n");
+        });
+
+        request = new Request("POST", "/" + index + "/_bulk?refresh=wait_for");
```
My [understanding](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-refresh.html) of the two is that `refresh=true` initiates a refresh right away, while `refresh=wait_for` will cause the response to only be given once the refresh is finished. I.e. `refresh=true` speeds up the test.

bpintea replied:
@astefan thanks for looking further into it. I've replaced `?refresh=wait_for` with `?refresh=true` (the last parameter being the one that applies) throughout the class.

bpintea commented on x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/JdbcAssert.java:
```diff
@@ -353,6 +352,89 @@ else if (type == Types.VARCHAR && actualObject instanceof TemporalAmount) {
+                }
+            }
+            if (actualObject instanceof Point) {
+                // geo points are loaded form doc values where they are stored as long-encoded values leading
```
I've removed the comment (but kept the "with-delta" comparison, I guess the fields also returns them from docvalues?)

astefan replied:
At the moment, no. Extraction from other sources other than _source is something for the future.

bpintea commented on x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/rest/RestSqlTestCase.java:
```diff
@@ -1169,4 +1192,167 @@ public static void assertResponse(Map<String, Object> expected, Map<String, Obje
+        }
+
+        Map<String, Object> map() {
+            return map == null ? null : new HashMap<>(map);
```
To allow the caller to modify what's returned and still fetch the original if needed. (Flexibility not really used now, tho.)

astefan replied:
If it's not used, I'd say let's keep it simple for the moment and when we'll need it, we'll add the flexibility. My 2c

matriv replied:
+1

bpintea replied:
I've updated the returns to the respective unmodifiable collections.

bpintea commented on x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/rest/RestSqlTestCase.java:
```diff
@@ -1169,4 +1192,167 @@ public static void assertResponse(Map<String, Object> expected, Map<String, Obje
+                if (val instanceof Map) {
+                    @SuppressWarnings("unchecked")
+                    Map<String, Object> innerMap = (Map<String, Object>) val;
+                    if (randomNonNegativeByte() < 60) {
```
I've added some comments here and there.
In this case `// assuming a linear distribution of the random non-negative byte values (0-255), [access this branch] roughly 1/4 of the time`

astefan replied:
There are some helper methods already existent that into the same ballpark idea: `rarely()`, `usually()`, `frequently()`.

bpintea replied:
Thanks. I've switched to `usually()`, which boost the probability to over 50% from ~25%, but should not make a difference testing-wise.

bpintea commented on x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/rest/RestSqlTestCase.java:
```diff
@@ -1169,4 +1192,167 @@ public static void assertResponse(Map<String, Object> expected, Map<String, Obje
+        private List<T> values;
+
+        // the depth excludes the leaf
+        ExhaustiveMultiPathMapper(int depth) {
```
Makes sense, I've refactored.
(The initial idea was to allow recalling one mapper instance with different value suppliers, but this is superfluous after all.)

matriv commented on x-pack/plugin/sql/jdbc/src/test/java/org/elasticsearch/xpack/sql/jdbc/JdbcPreparedStatementTests.java:
```diff
@@ -585,6 +586,13 @@ public void testThrownExceptionsWhenSettingByteArrayValuesToEsTypes() throws SQL
         assertEquals("Conversion from type [byte[]] to [INTERVAL_DAY_TO_MINUTE] not supported", sqle.getMessage());
     }
 
+    public void testThrownExceptionWhenSettingArrayValue() throws SQLException {
```
Thx!

bpintea committed:
Address review feedback

- refactor the way results are returned in mapper;
- reuse existing probability functions.
- remove redundant `refresh=wait_for`.

bpintea committed:
Merge remote-tracking branch 'upstream/feat/sql-multivalue' into feat/multivalue-clients

bpintea committed:
Test fix

- undo returning unmodifiable collection in mapper.

bpintea committed:
Merge remote-tracking branch 'upstream/feat/sql-multivalue' into feat/multivalue-clients

astefan reviewed (approved):
LGTM

matriv reviewed (approved):
LGTM, Nice work!

bpintea committed:
Merge remote-tracking branch 'upstream/feat/sql-multivalue' into feat/multivalue-clients

bpintea committed:
Adjust test

- adjust test after upstream ticket has been addressed.

bpintea committed:
Update feature-introducing version

Now that 7.13.0 is available, update the release target to that version.

costin commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/EsType.java:
```diff
@@ -48,7 +48,24 @@
     GEO_SHAPE(                ExtraTypes.GEOMETRY),
-    SHAPE(                    ExtraTypes.GEOMETRY);
+    SHAPE(                    ExtraTypes.GEOMETRY),
+    BOOLEAN_ARRAY(            Types.ARRAY),
```
Discussed with @bpintea to refactor this to use only a generic `ARRAY` type.

costin commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/JdbcArray.java:
```diff
@@ -0,0 +1,99 @@
+
+    @Override
+    public Object getArray(long index, int count, Map<String, Class<?>> map) throws SQLException {
+        if (map == null || map.isEmpty()) {
```
Small preference, it's better to do the validation and error checking upfront as oppose to last.
That is check if the map is non null and non-empty and throw the exception vs what the way it is right now.

bpintea replied:
Updated it.
However, I believe we might need to do [further updates](https://github.com/elastic/elasticsearch/blob/80b04338e2bcce0fb7b150641932119c91ed2a56/x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/JdbcResultSet.java#L447) to stay consistent.

costin commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/JdbcArray.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    @Override
+    public ResultSet getResultSet() throws SQLException {
```
A follow-up should implement this method - typically by returning a ResultSet that contains two columns - index and value.
A good deal of clients rely on this method since they're used to parsing `ResultSet` vs `getArray()`.
See `JdbcDatabaseMetaData#emptySet` for example of creating a dummy `ResultSet`.

bpintea replied:
ResultSet methods added.

costin commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/JdbcArray.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    @Override
+    public void free() throws SQLException {
```
Once this method is called, all other methods should return an exception which is not the case.
There are no side-effects however this is the mandatory behavior described by the spec:

```
 * After <code>free</code> has been called, any attempt to invoke a
 * method other than <code>free</code> will result in a <code>SQLException</code>
 * being thrown.  If <code>free</code> is called multiple times, the subsequent
 * calls to <code>free</code> are treated as a no-op.
```

bpintea replied:
Right, on another look, that phrasing is mandatory. Check added.

costin commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/TypeConverter.java:
```diff
@@ -263,6 +266,29 @@ static Object convert(Object v, EsType columnType, String typeString) throws SQL
                 }
             case IP:
                 return v.toString();
+            case BOOLEAN_ARRAY:
```
Needs to be improved

costin commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/JdbcArray.java:
```diff
@@ -0,0 +1,99 @@
+    private final EsType type;
+    private final List<?> values;
+
+    public JdbcArray(EsType type, List<?> values) {
```
Pass in the base type separately instead of determining it from the `EsType`

bpintea replied:
Changed to pass the base type.

costin commented on x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/JdbcAssert.java:
```diff
@@ -353,6 +352,87 @@ else if (type == Types.VARCHAR && actualObject instanceof TemporalAmount) {
         }
     }
 
+    private static Object extractObject(ResultSet resultSet, int column, boolean fromEs, Class<?> expectedColumnClass) throws SQLException {
```
A lot of work went in here.

costin commented on x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/rest/RestSqlTestCase.java:
```diff
@@ -1169,4 +1190,171 @@ public static void assertResponse(Map<String, Object> expected, Map<String, Obje
+     * Some of the paths are randomly multiplied:
+     *   {"a": [{"b": {"c": 1}}}, {"b": {"c": 1}}}], ...}
+     */
+    private static class ExhaustiveMultiPathMapper<T> {
```
And here...

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/session/VersionCompatibilityChecks.java:
```diff
@@ -0,0 +1,54 @@
+
+public final class VersionCompatibilityChecks {
+
+    public static final SqlVersion INTRODUCING_ARRAY_TYPES = SqlVersion.fromId(V_7_13_0.id);
```
I wonder if we need to remove `INTRODUCING_` since it will be the same for all new features. Then again, we need to differentiate from those that are being removed.

costin reviewed (commented):
Let a few comments. Great work!

bpintea commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/JdbcArray.java:
```diff
@@ -0,0 +1,99 @@
+
+import static org.elasticsearch.xpack.sql.jdbc.TypeUtils.baseType;
+
+public class JdbcArray implements Array {
```
Right, thanks, fixed.
